piet-cairo/src/text.rs

```diff
@@ -210,235 +357,149 @@ impl TextLayout for CairoTextLayout {
-        return HitTestPoint::default();
-    }
-
-    // then test the beginning and end (common cases)
-    if let Some(hit) = point_x_in_grapheme(point.x, &start_bounds) {
-        return hit;
-    }
-    if let Some(hit) = point_x_in_grapheme(point.x, &end_bounds) {
-        return hit;
-    }
+impl CairoTextLayout {
+    fn update_width(&mut self, new_width: impl Into<Option<f64>>) -> Result<(), Error> {
+        if let Some(new_width) = new_width.into() {
+            let pango_width = new_width * pango::SCALE as f64;
+            self.pango_layout.set_width(pango_width as i32);
+        } else {
+            /*
+             * NOTE: -1 is the default value, however `update_width` *could*
+             * be called any number of times with different values so we need
+             * to make sure to reset back to default whenever we get no width
+             */
+            self.pango_layout.set_width(-1);
+        }
 
-    // Now that we know it's not beginning or end, begin binary search.
-    // Iterative style
-    let mut left = start;
-    let mut right = end;
-    loop {
-        // pick halfway point
-        let middle = left + ((right - left) / 2);
-
-        let grapheme_bounds = match get_grapheme_boundaries(font, text, middle) {
-            Some(bounds) => bounds,
-            None => return HitTestPoint::default(),
-        };
+        let mut line_metrics = Vec::new();
+        let mut y_distance = 0.;
+        let mut widest_logical_width = 0;
+        for line_index in 0..self.pango_layout.get_line_count() {
+            if let Some(line) = self.pango_layout.get_line_readonly(line_index) {
+                /*
+                 * NOTE: These values are not currently exposed so we need to get them
+                 * manually. It's kinda sucky I know
+                 *
+                 * TODO(ForLoveOfCats): Submit a PR to gtk-rs to expose these values
+                 */
+                let (start_offset, end_offset) = unsafe {
+                    let raw_line = line.to_glib_none();
+
+                    let start_offset = (*raw_line.0).start_index as usize;
+                    let length = (*raw_line.0).length as usize;
+
+                    (start_offset, start_offset + length)
+                };
 
-        if let Some(hit) = point_x_in_grapheme(point.x, &grapheme_bounds) {
-            return hit;
-        }
+                //Pango likes to give us the line range *without* the newline char(s)
+                let end_offset = {
+                    let mut end_offset = end_offset;
+                    loop {
+                        let next_byte = self.text.as_bytes().get(end_offset);
+                        match next_byte {
+                            Some(b'\r') | Some(b'\n') => {
+                                end_offset += 1;
+                            }
+
+                            _ => break end_offset,
+                        }
+                    }
+                };
```

Comment:
We don't want to take *all* the paragraph separators; two in a row should be two separate lines. What we should do here is check just the end of the line, something like:

```rust
let end_offset = match self.text.as_bytes()[end_offset..] {
    &[b'\r', b'\n', ...] => end_offset + 2,
    &[b'\r', ...] | &[b'\n', ...] => end_offset + 1,
    _ => end_offset,
```

(I'm not sure if my range matching syntax is quite right, but something like this)

Reply:
I was intending on being a bit more graceful of handling stuff like out of order CRLF ect but blank lines totally slipped my mind.

Sidenote: I am constantly amazed by the expressiveness of pattern matching